### ISSUE TITLE
Comparison table: Number of model evaluations

### ISSUE THREAD

**Issue by JohannesBuchner** [mattpitkin/samplers-demo#31]
Feel free to close this if you are not interested, but I was wondering if you want to include the number of model evaluations in the summary table as well. This could be helpful to see whether the methods can be scaled up to slow likelihood functions. 

For Stan this is a bit difficult to quantify, but in python functions, you can use the following trick:

```
def mylikelihood(param):
    mylikelihood.counter += 1
    return something
mylikelihood.counter = 0

sampler = MySampler(mylikelihood)
sampler.run()
print(mylikelihood.counter)
```

The numbers could go into the "Nested sampling" table, for example.

In there it currently shows the "num. posterior samples", but these are pretty meaningless ... I can give you an arbitrary high number by resampling the weighted samples. The number of effective samples would be better here.

To make this issue at least somewhat useful, here is a typo: "Kullback-L**ie**bler (KL) diver**eg**nce" -> "Kullback-Leibler (KL) divergence"

**Comment by mattpitkin** [comment]
Thanks for the comments. I did start to add a likelihood counter based on #17, but then went with wall time for the reason that you give, i.e., its hard to do for PyJAGS/PySTAN. But, doing it for the Python nested sampling routines does make sense. I'll put this in my nest update.

Regarding showing the number of effective samples vs. the number of posterior samples, I'm not sure I understand. I would have assumed that the generating the posterior samples by resampling from the full sample set based on the weights does essentially give you the effective number of posterior samples if you don't allow duplication of points. Or is my understanding wrong here?

**Comment by JohannesBuchner** [comment]
You are currently drawing `len(weights)` rejection samples, so those samplers returning extra weight=0 entries will get more chance to draw. You would have to change this to draw a fixed number for all samplers. For example: 
```
N = len(np.unique(np.random.choice(len(weights), size=10000, p=weights / weights.sum())))
```

... or just use the ESS.

**Comment by mattpitkin** [comment]
Thanks. So, just to double check, is drawing the posterior samples the way I'm currently doing it still a "correct" way of getting unbiased posterior samples? Or, should the ESS be used to set the number of samples to be drawn? Or, is it just that calculating the ESS is just a fairer way of comparing samplers?

From looking at the UltraNest [`cornerplot`](https://johannesbuchner.github.io/UltraNest/_modules/ultranest/plot.html#cornerplot) functions it seems that the main difference with what I'm doing is that you mask out the initial very low weight values. So, would you also recommend doing that?

**Comment by JohannesBuchner** [comment]
> Thanks. So, just to double check, is drawing the posterior samples the way I'm currently doing it still a "correct" way of getting unbiased posterior samples? 

Yes, that is fine. Although ultranest already provides equally-weighted posteriors, which you could use directly (`results['samples']`).

> Or, should the ESS be used to set the number of samples to be drawn? 

That's one approach, but I think for some plotting routines oversampling the ESS by some factor and passing that in can be useful.
> Or, is it just that calculating the ESS is just a fairer way of comparing samplers?

Yes, that is what I was getting at, because in that case you do not introduce extra random noise either.

> From looking at the UltraNest [`cornerplot`](https://johannesbuchner.github.io/UltraNest/_modules/ultranest/plot.html#cornerplot) functions it seems that the main difference with what I'm doing is that you mask out the initial very low weight values. So, would you also recommend doing that?

This is for working around a corner plotting issue in that it makes the plot weird if you include extremely low-weight points.
